### ISSUE TITLE
Nextflow mounts /home into container, shadowing container home and causing command not found errors (exit 127)

### ISSUE THREAD

**Issue by mthartma** [epi2me-labs/wf-human-variation#108]
### Operating System

CentOS 7

### Workflow Version

v1.8.3-g93a02af

### Workflow Execution

Command line

### CLI command run

nextflow run epi2me-labs/wf-human-variation -w ${OUTPUT}/workspace -profile singularity --sv --snp --bam demo_data/demo.bam --bed demo_data/demo.bed --ref demo_data/demo.fasta --basecaller_cfg 'dna_r10.4.1_e8.2_400bps_hac@v4.1.0' --sample_name MY_SAMPLE --out_dir ${OUTPUT}

### Workflow Execution - CLI Execution Profile

None

### What happened?

I'm trying to validate the work flow on our system for a user by running your epi2me-labs/wf-human-variation demo and I have yet to get it to work. I'm using the [-profile singularity] as we don't run docker. It _looks_ like the container can't find the python binary. Just trying to figure out what is going wrong.

### Relevant log output

```shell
N E X T F L O W  ~  version 23.04.3
Launching `https://github.com/epi2me-labs/wf-human-variation` [deadly_heyrovsky] DSL2 - revision: 93a02af3ff [master]

||||||||||   _____ ____ ___ ____  __  __ _____      _       _
||||||||||  | ____|  _ \_ _|___ \|  \/  | ____|    | | __ _| |__  ___
|||||       |  _| | |_) | |  __) | |\/| |  _| _____| |/ _` | '_ \/ __|
|||||       | |___|  __/| | / __/| |  | | |__|_____| | (_| | |_) \__ \
||||||||||  |_____|_|  |___|_____|_|  |_|_____|    |_|\__,_|_.__/|___/
||||||||||  wf-human-variation v1.8.3-g93a02af
--------------------------------------------------------------------------------
Core Nextflow options
  revision       : master
  runName        : deadly_heyrovsky
  containerEngine: singularity
  container      : ontresearch/wf-human-variation:sha0800eade05e4cbb75d45421633c78c4f6320b2f6
  launchDir      : /home/groups/ruthm/mthartma/src/ticket
  workDir        : /home/groups/ruthm/mthartma/src/ticket/output/workspace
  projectDir     : /home/users/mthartma/.nextflow/assets/epi2me-labs/wf-human-variation
  userName       : mthartma
  profile        : singularity
  configFiles    : /home/users/mthartma/.nextflow/assets/epi2me-labs/wf-human-variation/nextflow.config

Workflow Options
  sv             : true
  snp            : true

Main options
  sample_name    : MY_SAMPLE
  bam            : demo_data/demo.bam
  ref            : demo_data/demo.fasta
  basecaller_cfg : dna_r10.4.1_e8.2_400bps_hac@v4.1.0
  bed            : demo_data/demo.bed

!! Only displaying parameters that differ from the pipeline defaults !!
--------------------------------------------------------------------------------
If you use epi2me-labs/wf-human-variation for your analysis please cite:

* The nf-core framework
  https://doi.org/10.1038/s41587-020-0439-x


--------------------------------------------------------------------------------
This is epi2me-labs/wf-human-variation v1.8.3-g93a02af.
--------------------------------------------------------------------------------
[-        ] process > bam_ingress:check_for_alignment   -
[-        ] process > bam_ingress:cram_to_bam           -
[-        ] process > bam_ingress:minimap2_alignment    -
executor >  local (4)
[-        ] process > bam_ingress:check_for_alignment   [  0%] 0 of 1
[-        ] process > bam_ingress:cram_to_bam           -
executor >  local (4)
[-        ] process > bam_ingress:check_for_alignment   [  0%] 0 of 1
[-        ] process > bam_ingress:cram_to_bam           -
[-        ] process > bam_ingress:minimap2_alignment    -
[-        ] process > getGenome                         -
[47/cfd872] process > cram_cache (1)                    [  0%] 0 of 1
[-        ] process > mosdepth_input                    -
[-        ] process > readStats                         -
[be/f78584] process > getVersions                       [  0%] 0 of 1
[-        ] process > getParams                         [  0%] 0 of 1
[-        ] process > get_region_coverage               -
[-        ] process > makeAlignmentReport               -
[-        ] process > failedQCReport                    -
[-        ] process > lookup_clair3_model               -
[-        ] process > snp:make_chunks                   -
[-        ] process > snp:pileup_variants               -
[-        ] process > snp:aggregate_pileup_variants     -
[-        ] process > snp:select_het_snps               -
[-        ] process > snp:phase_contig                  -
[-        ] process > snp:get_qual_filter               -
[-        ] process > snp:create_candidates             -
[-        ] process > snp:evaluate_candidates           -
[-        ] process > snp:aggregate_full_align_variants -
[-        ] process > snp:merge_pileup_and_full_vars    -
[-        ] process > snp:aggregate_all_variants        -
[-        ] process > sv:variantCall:filterBam          -
[-        ] process > sv:variantCall:sniffles2          -
[-        ] process > sv:variantCall:filterCalls        -
[-        ] process > sv:variantCall:sortVCF            -
[-        ] process > sv:variantCall:indexVCF           -
[-        ] process > sv:annotate_sv_vcf                -
[aa/476f56] process > sv:runReport:getVersions          [  0%] 0 of 1
[d7/139a29] process > sv:runReport:getParams            [  0%] 0 of 1
[-        ] process > sv:runReport:report               -
[-        ] process > output_sv                         -
executor >  local (5)
[-        ] process > bam_ingress:check_for_alignment   [  0%] 0 of 1
[-        ] process > bam_ingress:cram_to_bam           -
[-        ] process > bam_ingress:minimap2_alignment    -
[-        ] process > getGenome                         -
[47/cfd872] process > cram_cache (1)                    [  0%] 0 of 1
[-        ] process > mosdepth_input                    -
[-        ] process > readStats                         -
[be/f78584] process > getVersions                       [  0%] 0 of 1
[-        ] process > getParams                         [  0%] 0 of 1
[-        ] process > get_region_coverage               -
[-        ] process > makeAlignmentReport               -
[-        ] process > failedQCReport                    -
[-        ] process > lookup_clair3_model               -
[-        ] process > snp:make_chunks                   -
[-        ] process > snp:pileup_variants               -
[-        ] process > snp:aggregate_pileup_variants     -
[-        ] process > snp:select_het_snps               -
[-        ] process > snp:phase_contig                  -
[-        ] process > snp:get_qual_filter               -
[-        ] process > snp:create_candidates             -
[-        ] process > snp:evaluate_candidates           -
[-        ] process > snp:aggregate_full_align_variants -
[-        ] process > snp:merge_pileup_and_full_vars    -
[-        ] process > snp:aggregate_all_variants        -
[-        ] process > sv:variantCall:filterBam          -
[-        ] process > sv:variantCall:sniffles2          -
[-        ] process > sv:variantCall:filterCalls        -
[-        ] process > sv:variantCall:sortVCF            -
[-        ] process > sv:variantCall:indexVCF           -
[-        ] process > sv:annotate_sv_vcf                -
[aa/476f56] process > sv:runReport:getVersions          [100%] 1 of 1, failed: 1 ✘
[d7/139a29] process > sv:runReport:getParams            [  0%] 0 of 1
[-        ] process > sv:runReport:report               -
[-        ] process > output_sv                         -
executor >  local (5)
[-        ] process > bam_ingress:check_for_alignment   [  0%] 0 of 1
[-        ] process > bam_ingress:cram_to_bam           -
[-        ] process > bam_ingress:minimap2_alignment    -
[-        ] process > getGenome                         -
[47/cfd872] process > cram_cache (1)                    [  0%] 0 of 1
[-        ] process > mosdepth_input                    -
[-        ] process > readStats                         -
[be/f78584] process > getVersions                       [  0%] 0 of 1
[-        ] process > getParams                         [  0%] 0 of 1
[-        ] process > get_region_coverage               -
[-        ] process > makeAlignmentReport               -
[-        ] process > failedQCReport                    -
[-        ] process > lookup_clair3_model               -
[-        ] process > snp:make_chunks                   -
[-        ] process > snp:pileup_variants               -
[-        ] process > snp:aggregate_pileup_variants     -
[-        ] process > snp:select_het_snps               -
[-        ] process > snp:phase_contig                  -
[-        ] process > snp:get_qual_filter               -
[-        ] process > snp:create_candidates             -
[-        ] process > snp:evaluate_candidates           -
[-        ] process > snp:aggregate_full_align_variants -
[-        ] process > snp:merge_pileup_and_full_vars    -
[-        ] process > snp:aggregate_all_variants        -
[-        ] process > sv:variantCall:filterBam          -
[-        ] process > sv:variantCall:sniffles2          -
[-        ] process > sv:variantCall:filterCalls        -
[-        ] process > sv:variantCall:sortVCF            -
[-        ] process > sv:variantCall:indexVCF           -
[-        ] process > sv:annotate_sv_vcf                -
[aa/476f56] process > sv:runReport:getVersions          [100%] 1 of 1, failed: 1 ✘
[d7/139a29] process > sv:runReport:getParams            [  0%] 0 of 1
[-        ] process > sv:runReport:report               -
[-        ] process > output_sv                         -
[-        ] process > refine_with_sv                    -
[-        ] process > bed_filter                        -
[-        ] process > annotate_snp_vcf                  -
[-        ] process > vcfStats                          -
[-        ] process > report_snp:getVersions            -
[-        ] process > report_snp:getParams              -
[-        ] process > report_snp:makeReport             -
[-        ] process > output_snp                        -
[-        ] process > configure_jbrowse                 -
[c5/2c8e7b] process > publish_artifact (1)              [  0%] 0 of 1
Pulling Singularity image docker://ontresearch/wf-human-variation:sha0800eade05e4cbb75d45421633c78c4f6320b2f6 [cache /home/groups/ruthm/mthartma/src/ticket/ontresearch-wf-human-variation-sha0800eade05e4cbb75d45421633c78c4f6320b2f6.img]
Pulling Singularity image docker://ontresearch/wf-human-variation-snp:sha0d7e7e8e8207d9d23fdf50a34ceb577da364373e [cache /home/groups/ruthm/mthartma/src/ticket/ontresearch-wf-human-variation-snp-sha0d7e7e8e8207d9d23fdf50a34ceb577da364373e.img]
Pulling Singularity image docker://ontresearch/wf-human-variation-sv:shabc3ac908a14705f248cdf49f218956ec33e93ef9 [cache /home/groups/ruthm/mthartma/src/ticket/ontresearch-wf-human-variation-sv-shabc3ac908a14705f248cdf49f218956ec33e93ef9.img]
ERROR ~ Error executing process > 'sv:runReport:getVersions'

Caused by:
  Process `sv:runReport:getVersions` terminated with an error exit status (127)

Command executed:

  trap '' PIPE # suppress SIGPIPE without interfering with pipefail
  python -c "import pysam; print(f'pysam,{pysam.__version__}')" >> versions.txt
  truvari version | sed 's/ /,/' >> versions.txt
  fastcat --version | sed 's/^/fastcat,/' >> versions.txt
  sniffles --version | head -n 1 | sed 's/ Version //' >> versions.txt
  bcftools --version | head -n 1 | sed 's/ /,/' >> versions.txt
  samtools --version | head -n 1 | sed 's/ /,/' >> versions.txt
  minimap2 --version | head -n 1 | sed 's/^/minimap2,/' >> versions.txt
  echo `seqtk 2>&1 | head -n 3 | tail -n 1 | cut -d ':' -f 2 | sed 's/ /seqtk,/'` >> versions.txt

Command exit status:
  127

Command output:
  (empty)

Command error:
  INFO:    Environment variable SINGULARITYENV_TMPDIR is set, but APPTAINERENV_TMPDIR is preferred
  INFO:    Environment variable SINGULARITYENV_NXF_DEBUG is set, but APPTAINERENV_NXF_DEBUG is preferred
  INFO:    underlay of /etc/localtime required more than 50 (78) bind mounts
  .command.sh: line 3: python: command not found
  ps: error while loading shared libraries: liblz4.so.1: cannot open shared object file: No such file or directory

Work dir:
  /home/groups/ruthm/mthartma/src/ticket/output/workspace/aa/476f56e3f1bb56df302e3ed6240609

Tip: view the complete command output by changing to the process work dir and entering the command `cat .command.out`

 -- Check '.nextflow.log' file for details
WARN: Killing running tasks (4)
```

**Comment by cjw85** [comment]
Hi @mthartma,

Have you managed to resolve this? Something doesn't seem quite right with the singularity environment or images that have been created (from our docker images). @SamStudio8 is more knowlegeable that me with singularity.

**Comment by mthartma** [comment]
Hi Chris,

I was waiting on more info/direction from you guys. According to my internal ticket the user was also having issues running the Demo.

Thanks,

Mike

**Comment by SamStudio8** [comment]
Hm. I am not _entirely_ sure what to make of this one. It is possible that this error is actually a permissions error masquerading as a command not found error, and the user does not have permissions to run the `python` binary included with the container. Let's try and confirm our suspicions by asking the container for its PATH and whether it can access the bin directory that we ship Python in:

```
singularity exec /home/groups/ruthm/mthartma/src/ticket/ontresearch-wf-human-variation-sv-shabc3ac908a14705f248cdf49f218956ec33e93ef9.img bash -c 'echo $PATH; namei -l /home/epi2melabs/conda/bin/python; which python'
```

@mthartma Before we get too in the weeds, can I just check if you have any options that you generally apply to `singularity exec` when running containers for other tasks? I note you're on CentOS which has required some special configuration to correctly bind mounts for Docker for other users.

**Comment by ffavero** [comment]
I have the same issue, also on a CentOS HPC.

running singularity seems fine

```
[frafav@g-03-c0178 epi2me-labs]$ singularity exec wf-human-variation/cachedir/ontresearch-wf-human-variation-sv-shabc3ac908a14705f248cdf49f218956ec33e93ef9.img bas
h -c 'echo $PATH; namei -l /home/epi2melabs/conda/bin/python; which python'
INFO:    Converting SIF file to temporary sandbox...
WARNING: underlay of /etc/localtime required more than 50 (78) bind mounts
/home/epi2melabs/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
f: /home/epi2melabs/conda/bin/python
drwxr-xr-x frafav s093829 /
drwxr-xr-x frafav s093829 home
drwxrwxr-x frafav s093829 epi2melabs
drwxrwxr-x frafav s093829 conda
drwxrwxr-x frafav s093829 bin
lrwxrwxrwx frafav s093829 python -> python3.8
-rwxr-xr-x frafav s093829   python3.8
/home/epi2melabs/conda/bin/python
INFO:    Cleaning up image...
```

One thing I notived is that running different version of singularity might lead to different error, eg with 3.10.5 version I have a very similar error from OP, when trying with singularity 4.0.0 I have all sort of error related to image conversion. I wonder if there is a specific singularity version we should use

**Comment by ffavero** [comment]
I amend my previous statement, my images errors were related to something else in my environment.
With various singularity version I now have always the same error

```
ERROR ~ Error executing process > 'sv:runReport:getVersions'

Caused by:
  Process `sv:runReport:getVersions` terminated with an error exit status (127)

Command executed:

  trap '' PIPE # suppress SIGPIPE without interfering with pipefail
  python -c "import pysam; print(f'pysam,{pysam.__version__}')" >> versions.txt
  truvari version | sed 's/ /,/' >> versions.txt
  fastcat --version | sed 's/^/fastcat,/' >> versions.txt
  sniffles --version | head -n 1 | sed 's/ Version //' >> versions.txt
  bcftools --version | head -n 1 | sed 's/ /,/' >> versions.txt
  samtools --version | head -n 1 | sed 's/ /,/' >> versions.txt
  minimap2 --version | head -n 1 | sed 's/^/minimap2,/' >> versions.txt
  echo `seqtk 2>&1 | head -n 3 | tail -n 1 | cut -d ':' -f 2 | sed 's/ /seqtk,/'` >> versions.txt

Command exit status:
  127

Command output:
  (empty)

Command error:
  INFO:    Converting SIF file to temporary sandbox...
  WARNING: underlay of /etc/localtime required more than 50 (78) bind mounts
  .command.sh: line 3: python: command not found
  INFO:    Cleaning up image...

Work dir:
  /home/projects/cu_10027/data/pipelines/nextflow/epi2me-labs/output/workspace/9f/a77fa35af8c026310c0f5cf8b94445

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line

 -- Check '.nextflow.log' file for details

```

**Comment by mthartma** [comment]
I was able to do the check and it seems fine:

$ singularity exec /home/groups/ruthm/mthartma/src/ticket/ontresearch-wf-human-variation-sv-shabc3ac908a14705f248cdf49f218956ec33e93ef9.img bash -c 'echo $PATH; namei -l /home/epi2melabs/conda/bin/python; which python'
INFO:    underlay of /etc/localtime required more than 50 (78) bind mounts
/home/epi2melabs/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
f: /home/epi2melabs/conda/bin/python
drwxr-xr-x mthartma ruthm /
drwxr-xr-x mthartma ruthm home
drwsrwsr-x mthartma ruthm epi2melabs
drwsrwsr-x mthartma ruthm conda
drwsrwsr-x mthartma ruthm bin
lrwxrwxrwx mthartma ruthm python -> python3.8
-rwxrwxr-x mthartma ruthm   python3.8
/home/epi2melabs/conda/bin/python

**Comment by SamStudio8** [comment]
@mthartma @ffavero Thanks both, just for my own sanity: does `python` actually run when executed this way?

```
singularity exec /path/to.img bash -c 'echo $PATH; namei -l /home/epi2melabs/conda/bin/python; which python; python -c "print(\"hello world\")"'
```

**Comment by mthartma** [comment]
Sanity check confirmed:

[mthartma] (job 35379246) $ singularity exec /home/groups/ruthm/mthartma/src/ticket/ontresearch-wf-human-variation-sv-shabc3ac908a14705f248cdf49f218956ec33e93ef9.img bash -c 'echo $PATH; namei -l /home/epi2melabs/conda/bin/python; which python; python -c "print(\"hello world\")"'
INFO:    underlay of /etc/localtime required more than 50 (78) bind mounts
/home/epi2melabs/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
f: /home/epi2melabs/conda/bin/python
drwxr-xr-x mthartma ruthm /
drwxr-xr-x mthartma ruthm home
drwsrwsr-x mthartma ruthm epi2melabs
drwsrwsr-x mthartma ruthm conda
drwsrwsr-x mthartma ruthm bin
lrwxrwxrwx mthartma ruthm python -> python3.8
-rwxrwxr-x mthartma ruthm   python3.8
/home/epi2melabs/conda/bin/python
hello world
[mthartma] (job 35379246) $

**Comment by ffavero** [comment]
> @mthartma @ffavero Thanks both, just for my own sanity: does `python` actually run when executed this way?
> 
> ```
> singularity exec /path/to.img bash -c 'echo $PATH; namei -l /home/epi2melabs/conda/bin/python; which python; python -c "print(\"hello world\")"'
> ```

yes, it does, I was even able to use the image interactively with a python console

**Comment by ffavero** [comment]
sorry, pasting the actual proof that this works

```bash
[frafav@g-03-c0178 epi2me-labs]$ singularity exec  wf-human-variation/cachedir/ontresearch-wf-human-variation-sv-shabc3ac908a14705f248cdf49f218956ec33e93ef9.img bash -c 'echo $PATH; namei -l /home/epi2melabs/conda/bin/python; which python; python -c "print(\"hello world\")"'
INFO:    Converting SIF file to temporary sandbox...
WARNING: underlay of /etc/localtime required more than 50 (78) bind mounts
/home/epi2melabs/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
f: /home/epi2melabs/conda/bin/python
drwxr-xr-x frafav s093829 /
drwxr-xr-x frafav s093829 home
drwxrwxr-x frafav s093829 epi2melabs
drwxrwxr-x frafav s093829 conda
drwxrwxr-x frafav s093829 bin
lrwxrwxrwx frafav s093829 python -> python3.8
-rwxr-xr-x frafav s093829   python3.8
/home/epi2melabs/conda/bin/python
hello world
INFO:    Cleaning up image...
```

**Comment by ffavero** [comment]
I trying to set libraryDis as well as cacheDir and use a specific singularity tempdir. I have always the same outcome, sometime the job that fails is a different one (see log pasted below), but it's probably a concurrency matter 

```
ERROR ~ Error executing process > 'report_snp:getVersions'
                                                                                                                                                                   
Caused by:
  Process `report_snp:getVersions` terminated with an error exit status (127)

Command executed:

  run_clair3.sh --version | sed 's/ /,/' >> versions.txt

Command exit status:
  127

Command output:
  (empty)

Command error:
  INFO:    Converting SIF file to temporary sandbox...
  WARNING: underlay of /etc/localtime required more than 50 (78) bind mounts
  .command.sh: line 2: run_clair3.sh: command not found
  INFO:    Cleaning up image...
  ERROR:   failed to delete container image tempDir /home/projects/cu_10027/scratch/rootfs-653283536: unlinkat /home/projects/cu_10027/scratch/rootfs-653283536/roo
t/etc: directory not empty

Work dir:
  /home/projects/cu_10027/data/pipelines/nextflow/epi2me-labs/output/workspace/1d/15c6711061c8f893f645f4607a8a22

Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`

 -- Check '.nextflow.log' file for details
```

when adding a line in `.command.sh` like this:

```bash
#!/bin/bash -euo pipefail
trap '' PIPE # suppress SIGPIPE without interfering with pipefail
echo $PATH
ls /home
python -c "import pysam; print(f'pysam,{pysam.__version__}')" >> versions.txt
truvari version | sed 's/ /,/' >> versions.txt
fastcat --version | sed 's/^/fastcat,/' >> versions.txt
sniffles --version | head -n 1 | sed 's/ Version //' >> versions.txt
bcftools --version | head -n 1 | sed 's/ /,/' >> versions.txt
samtools --version | head -n 1 | sed 's/ /,/' >> versions.txt
minimap2 --version | head -n 1 | sed 's/^/minimap2,/' >> versions.txt
echo `seqtk 2>&1 | head -n 3 | tail -n 1 | cut -d ':' -f 2 | sed 's/ /seqtk,/'` >> versions.txt
```

I get the "missing python" error, and apparently it didn't mounted the singularity image correctly, as the home `epi2melabs` folder is missing

```
[frafav@g-02-c0116 d5edce0e035e1177b556ecd1d0a6df]$ bash .command.run
INFO:    Converting SIF file to temporary sandbox...
WARNING: underlay of /etc/localtime required more than 50 (78) bind mounts
/home/epi2melabs/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/home/people/frafav/.nextflow/assets/epi2me-labs/wf-human-variation/bin
databases
people
projects
/home/projects/cu_10027/data/pipelines/nextflow/epi2me-labs/output/workspace/45/d5edce0e035e1177b556ecd1d0a6df/.command.sh: line 5: python: command not found
INFO:    Cleaning up image...
```

**Comment by SamStudio8** [comment]
@ffavero Thanks for the debugging here! I meant to get back to you both about this after my last message, as I had a feeling if the image was working with a plain call to exec, it must be something to do with the bindings added by Nextflow. I had a sinking feeling it was mounting `/home` and obscuring the image, thanks for confirming.

If you can see `-B /home` in the `.command.run`, it is possible that Nextflow has added this as the longest common path of your inputs. In the example initially posted by @mthartma, I can see that the launchDir has a path from `/home/groups/...` but the Nextflow workflow has been installed to `/home/users/...`.

@ffavero Could you share the top few lines of the Nextflow stdout that show the paths for the launchDir, inputs and projectDir. I wonder if you have a similar set up.

We are actually planning to move the host data out of `home/` in our images to try and avoid polluting home in this way, but obviously that doesn't help you right now.

**Comment by SamStudio8** [comment]
Indeed, looks like Nextflow mounts the longest common path of the input files and the bin dirs (ie. the Nextflow assets):
```
./modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy:532:            builder.addMountForInputs(inputFiles)
./modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy:535:            builder.addMounts(binDirs)
```
```
./modules/nextflow/src/main/groovy/nextflow/container/ContainerBuilder.groovy:275:        // find the longest commons paths and mount only them
./modules/nextflow/src/main/groovy/nextflow/container/ContainerBuilder.groovy:280:        final paths = trie.longest()
```

**Comment by ffavero** [comment]
> @ffavero Thanks for the debugging here! I meant to get back to you both about this after my last message, as I had a feeling if the image was working with a plain call to exec, it must be something to do with the bindings added by Nextflow. I had a sinking feeling it was mounting `/home` and obscuring the image, thanks for confirming.
> 
> If you can see `-B /home` in the `.command.run`, it is possible that Nextflow has added this as the longest common path of your inputs. In the example initially posted by @mthartma, I can see that the launchDir has a path from `/home/groups/...` but the Nextflow workflow has been installed to `/home/users/...`.
> 
> @ffavero Could you share the top few lines of the Nextflow stdout that show the paths for the launchDir, inputs and projectDir. I wonder if you have a similar set up.
> 
> We are actually planning to move the containerised data out of `/home/epi2melabs` in our images to try and avoid polluting home in this way (and adhering to the FHS a bit better!), but obviously that doesn't help you right now.

I don’t have a terminal on hand to show a log, but I can definitely confirm a similar issue. The assets were in `/home/people/…` while the data are in `/home/projects/…`. However I have now changed the asset folder to be in projects as well, I have to test if the automont works doing so

**Comment by SamStudio8** [comment]
@ffavero Did you have any luck?
@mthartma I have enough confidence from what I've found above to suggest that the workflows will be able to run if you are able to install the workflow assets to elsewhere from `/home/...` in your directory hierarchy (`/home/groups/` would be acceptable as the longest common path of `/home/groups/` would not shadow `/home/` in that case). You can change the location of the assets dir by setting an alternative path for the Nextflow home with the `NXF_HOME` environment variable.

**Comment by ffavero** [comment]
> @ffavero Did you have any luck? @mthartma I have enough confidence from what I've found above to suggest that the workflows will be able to run if you are able to install the workflow assets to elsewhere from `/home/...` in your directory hierarchy (`/home/groups/` would be acceptable as the longest common path of `/home/groups/` would not shadow `/home/` in that case). You can change the location of the assets dir by setting an alternative path for the Nextflow home with the `NXF_HOME` environment variable.

Yes! indeed, I swithced back autoMount and set the NXF_HOME somewhere else in /home/projects, removed all my previous custom bindings from the config, and it worked!

I made a pbs/singularity profile and it works just fine... now onto testing with some real data :D

Thanks @SamStudio8

**Comment by SamStudio8** [comment]
Excellent news. I'll wait hear back that @mthartma has been able to get up and running before closing, but it looks like mystery solved.

**Comment by ffavero** [comment]
ho no, I had another error

```
N E X T F L O W  ~  version 23.04.3
Launching `https://github.com/epi2me-labs/wf-human-variation` [sad_leakey] DSL2 - revision: 93a02af3ff [master]

||||||||||   _____ ____ ___ ____  __  __ _____      _       _
||||||||||  | ____|  _ \_ _|___ \|  \/  | ____|    | | __ _| |__  ___
|||||       |  _| | |_) | |  __) | |\/| |  _| _____| |/ _` | '_ \/ __|
|||||       | |___|  __/| | / __/| |  | | |__|_____| | (_| | |_) \__ \
||||||||||  |_____|_|  |___|_____|_|  |_|_____|    |_|\__,_|_.__/|___/
||||||||||  wf-human-variation v1.8.3-g93a02af
--------------------------------------------------------------------------------
Core Nextflow options
  revision       : master
  runName        : sad_leakey
  containerEngine: singularity
  container      : ontresearch/wf-human-variation:sha0800eade05e4cbb75d45421633c78c4f6320b2f6
  launchDir      : /home/projects/cu_10027/data/pipelines/nextflow/epi2me-labs
  workDir        : /home/projects/cu_10027/data/pipelines/nextflow/epi2me-labs/output/workspace
  projectDir     : /home/projects/cu_10027/people/frafav/home_symlinks/nextflow/assets/epi2me-labs/wf-human-variation
  userName       : frafav
  profile        : singularity
  configFiles    : /home/projects/cu_10027/people/frafav/home_symlinks/nextflow/assets/epi2me-labs/wf-human-variation/nextflow.config, /home/projects/cu_10027/data/pipelines/nextflow/epi2me-labs/wf-human-variation/nextflow.config

Workflow Options
  sv             : true
  snp            : true

Main options
  sample_name    : MY_SAMPLE
  bam            : /home/projects/cu_10027/data/pipelines/nextflow/epi2me-labs/demo_data/demo.bam
  ref            : /home/projects/cu_10027/data/pipelines/nextflow/epi2me-labs/demo_data/demo.fasta
  basecaller_cfg : dna_r10.4.1_e8.2_400bps_hac@v4.1.0
  bed            : /home/projects/cu_10027/data/pipelines/nextflow/epi2me-labs/demo_data/demo.bed

!! Only displaying parameters that differ from the pipeline defaults !!
--------------------------------------------------------------------------------
If you use epi2me-labs/wf-human-variation for your analysis please cite:

* The nf-core framework
  https://doi.org/10.1038/s41587-020-0439-x
```

this is the error

```
ERROR ~ Error executing process > 'snp:aggregate_all_variants (1)'

Caused by:
  Process `snp:aggregate_all_variants (1)` terminated with an error exit status (2)

Command executed:

  prefix="merge"
  phase_vcf=false
  if [[ $phase_vcf == "true" ]]; then
      prefix="phased"
  fi
  ls merge_output/*.vcf.gz | parallel --jobs 4 "bgzip -d {}"

  pypy $(which clair3.py) SortVcf             --input_dir merge_output             --vcf_fn_prefix $prefix             --output_fn MY_SAMPLE.wf_snp.vcf             --sampleName MY_SAMPLE             --ref_fn demo.fasta             --cmd_fn CMD             --contigs_fn CONTIGS

  if [ "$( bgzip -fdc MY_SAMPLE.wf_snp.vcf.gz | grep -v '#' | wc -l )" -eq 0 ]; then
      echo "[INFO] Exit in all contigs variant merging"
      exit 0
  fi

  # TODO: this could be a separate process
  if [ "false" == "true" ]; then
      pypy $(which clair3.py) SortVcf                 --input_dir merge_outputs_gvcf                 --vcf_fn_prefix merge                 --vcf_fn_suffix .gvcf                 --output_fn tmp.gvcf                 --sampleName MY_SAMPLE                 --ref_fn demo.fasta                 --cmd_fn CMD                 --contigs_fn CONTIGS

          # Reheading samples named "SAMPLE" to params.sample_name. If no
          echo "SAMPLE" "MY_SAMPLE" > rename.txt
          bcftools reheader -s rename.txt tmp.gvcf.gz > MY_SAMPLE.wf_snp.gvcf.gz
          bcftools index -t MY_SAMPLE.wf_snp.gvcf.gz && rm tmp.gvcf.gz rename.txt
  fi

  echo "[INFO] Finish calling, output file: merge_output.vcf.gz"

Command exit status:
  2

Command output:
  (empty)

Command error:
  INFO:    Converting SIF file to temporary sandbox...
  WARNING: underlay of /etc/localtime required more than 50 (78) bind mounts
  Error in tempfile() using template /scratch/47130881/parXXXXX.par: Parent directory (/scratch/47130881/) does not exist at /home/epi2melabs/conda/bin/parallel line 5150.
  INFO:    Cleaning up image...

Work dir:
  /home/projects/cu_10027/data/pipelines/nextflow/epi2me-labs/output/workspace/bb/11684829dae0abab9c7f4b220545e1

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line

 -- Check '.nextflow.log' file for details
```

This even after few retries with the `-resume` flag. I was calling victory because it made almost all the workflow steps before this error

**Comment by mthartma** [comment]
Hi @SamStudio8 , I'll give this a try later today or early tomorrow.
Thanks for all the help

Mike

**Comment by SamStudio8** [comment]
@ffavero Weird - Did you have a TMPDIR set to this `/scratch/...` path in the environment where you launched Nextflow? Looks like some sort of environment information might have bled into the containers?

**Comment by mthartma** [comment]
not so great luck...

[mthartma@sh02-01n58 /home/groups/ruthm/mthartma/src/ticket] (job 35809621) $ export SINGULARITYENV_TMPDIR=/home/groups/ruthm/mthartma/src/ticket
[mthartma@sh02-01n58 /home/groups/ruthm/mthartma/src/ticket] (job 35809621) $ export APPTAINER_TMPDIR=/home/groups/ruthm/mthartma/src/ticket
[mthartma@sh02-01n58 /home/groups/ruthm/mthartma/src/ticket] (job 35809621) $ export SINGULARITYENV_NXF_DEBUG=/home/groups/ruthm/mthartma/src/ticket
[mthartma@sh02-01n58 /home/groups/ruthm/mthartma/src/ticket] (job 35809621) $ export APPTAINERENV_NXF_DEBUG=/home/groups/ruthm/mthartma/src/ticket
[mthartma@sh02-01n58 /home/groups/ruthm/mthartma/src/ticket] (job 35809621) $ nextflow run epi2me-labs/wf-human-variation -w ${OUTPUT}/workspace -profile singularity --snp --sv --bam demo_data/demo.bam --bed demo_data/demo.bed --ref demo_data/demo.fasta --basecaller_cfg 'dna_r10.4.1_e8.2_400bps_hac@v4.1.0' --sample_name MY_SAMPLE --out_dir ${OUTPUT}
N E X T F L O W  ~  version 23.04.3
Launching `https://github.com/epi2me-labs/wf-human-variation` [golden_faggin] DSL2 - revision: 93a02af3ff [master]

||||||||||   _____ ____ ___ ____  __  __ _____      _       _
||||||||||  | ____|  _ \_ _|___ \|  \/  | ____|    | | __ _| |__  ___
|||||       |  _| | |_) | |  __) | |\/| |  _| _____| |/ _` | '_ \/ __|
|||||       | |___|  __/| | / __/| |  | | |__|_____| | (_| | |_) \__ \
||||||||||  |_____|_|  |___|_____|_|  |_|_____|    |_|\__,_|_.__/|___/
||||||||||  wf-human-variation v1.8.3-g93a02af
--------------------------------------------------------------------------------
Core Nextflow options
  revision       : master
  runName        : golden_faggin
  containerEngine: singularity
  container      : ontresearch/wf-human-variation:sha0800eade05e4cbb75d45421633c78c4f6320b2f6
  launchDir      : /home/groups/ruthm/mthartma/src/ticket
  workDir        : /home/groups/ruthm/mthartma/src/ticket/output/workspace
  projectDir     : /home/users/mthartma/.nextflow/assets/epi2me-labs/wf-human-variation
  userName       : mthartma
  profile        : singularity
  configFiles    : /home/users/mthartma/.nextflow/assets/epi2me-labs/wf-human-variation/nextflow.config

Workflow Options
  sv             : true
  snp            : true

Main options
  sample_name    : MY_SAMPLE
  bam            : demo_data/demo.bam
  ref            : demo_data/demo.fasta
  basecaller_cfg : dna_r10.4.1_e8.2_400bps_hac@v4.1.0
  bed            : demo_data/demo.bed

!! Only displaying parameters that differ from the pipeline defaults !!
--------------------------------------------------------------------------------
If you use epi2me-labs/wf-human-variation for your analysis please cite:

* The nf-core framework
  https://doi.org/10.1038/s41587-020-0439-x


--------------------------------------------------------------------------------
This is epi2me-labs/wf-human-variation v1.8.3-g93a02af.
--------------------------------------------------------------------------------
[-        ] process > bam_ingress:check_for_alignment   [  0%] 0 of 1
executor >  local (1)
executor >  local (2)
[-        ] process > bam_ingress:check_for_alignment   [  0%] 0 of 1
[-        ] process > bam_ingress:cram_to_bam           -
[-        ] process > bam_ingress:minimap2_alignment    -
[-        ] process > getGenome                         -
[95/0246b8] process > cram_cache (1)                    [  0%] 0 of 1
[-        ] process > mosdepth_input                    -
[-        ] process > readStats                         -
[60/2de55e] process > getVersions                       [  0%] 0 of 1
[-        ] process > getParams                         [  0%] 0 of 1
[-        ] process > get_region_coverage               -
[-        ] process > makeAlignmentReport               -
[-        ] process > failedQCReport                    -
[-        ] process > lookup_clair3_model               -
[-        ] process > snp:make_chunks                   -
[-        ] process > snp:pileup_variants               -
[-        ] process > snp:aggregate_pileup_variants     -
[-        ] process > snp:select_het_snps               -
[-        ] process > snp:phase_contig                  -
[-        ] process > snp:get_qual_filter               -
executor >  local (2)
[-        ] process > bam_ingress:check_for_alignment   [  0%] 0 of 1
[-        ] process > bam_ingress:cram_to_bam           -
[-        ] process > bam_ingress:minimap2_alignment    -
[-        ] process > getGenome                         -
[95/0246b8] process > cram_cache (1)                    [100%] 1 of 1, failed: 1 ✘
[-        ] process > mosdepth_input                    -
[-        ] process > readStats                         -
[60/2de55e] process > getVersions                       [  0%] 0 of 1
[-        ] process > getParams                         [  0%] 0 of 1
[-        ] process > get_region_coverage               -
[-        ] process > makeAlignmentReport               -
[-        ] process > failedQCReport                    -
[-        ] process > lookup_clair3_model               -
[-        ] process > snp:make_chunks                   -
[-        ] process > snp:pileup_variants               -
[-        ] process > snp:aggregate_pileup_variants     -
[-        ] process > snp:select_het_snps               -
[-        ] process > snp:phase_contig                  -
[-        ] process > snp:get_qual_filter               -
executor >  local (2)
[-        ] process > bam_ingress:check_for_alignment   [  0%] 0 of 1
[-        ] process > bam_ingress:cram_to_bam           -
[-        ] process > bam_ingress:minimap2_alignment    -
[-        ] process > getGenome                         -
[95/0246b8] process > cram_cache (1)                    [100%] 1 of 1, failed: 1 ✘
[-        ] process > mosdepth_input                    -
[-        ] process > readStats                         -
[60/2de55e] process > getVersions                       [  0%] 0 of 1
[-        ] process > getParams                         [  0%] 0 of 1
[-        ] process > get_region_coverage               -
[-        ] process > makeAlignmentReport               -
[-        ] process > failedQCReport                    -
[-        ] process > lookup_clair3_model               -
[-        ] process > snp:make_chunks                   -
[-        ] process > snp:pileup_variants               -
[-        ] process > snp:aggregate_pileup_variants     -
[-        ] process > snp:select_het_snps               -
[-        ] process > snp:phase_contig                  -
[-        ] process > snp:get_qual_filter               -
[-        ] process > snp:create_candidates             -
[-        ] process > snp:evaluate_candidates           -
[-        ] process > snp:aggregate_full_align_variants -
[-        ] process > snp:merge_pileup_and_full_vars    -
[-        ] process > snp:aggregate_all_variants        -
[-        ] process > sv:variantCall:filterBam          -
[-        ] process > sv:variantCall:sniffles2          -
[-        ] process > sv:variantCall:filterCalls        -
[-        ] process > sv:variantCall:sortVCF            -
[-        ] process > sv:variantCall:indexVCF           -
[-        ] process > sv:annotate_sv_vcf                -
[-        ] process > sv:runReport:getVersions          [  0%] 0 of 1
[-        ] process > sv:runReport:getParams            [  0%] 0 of 1
[-        ] process > sv:runReport:report               -
[-        ] process > output_sv                         [  0%] 0 of 1
[-        ] process > refine_with_sv                    -
[-        ] process > bed_filter                        -
[-        ] process > annotate_snp_vcf                  -
[-        ] process > vcfStats                          -
[-        ] process > report_snp:getVersions            -
[-        ] process > report_snp:getParams              -
[-        ] process > report_snp:makeReport             -
[-        ] process > output_snp                        -
[-        ] process > configure_jbrowse                 -
[-        ] process > publish_artifact                  [  0%] 0 of 1
Pulling Singularity image docker://ontresearch/wf-human-variation-snp:sha0d7e7e8e8207d9d23fdf50a34ceb577da364373e [cache /home/groups/ruthm/mthartma/src/ticket/ontresearch-wf-human-variation-snp-sha0d7e7e8e8207d9d23fdf50a34ceb577da364373e.img]
ERROR ~ Error executing process > 'cram_cache (1)'

Caused by:
  Process `cram_cache (1)` terminated with an error exit status (127)

Command executed:

  # Invoke from binary installed to container PATH
  seq_cache_populate.pl -root ref_cache/ demo.fasta
  REF_PATH="ref_cache/%2s/%2s/%s"

Command exit status:
  127

Command output:
  (empty)

Command error:
  INFO:    Environment variable SINGULARITYENV_TMPDIR is set, but APPTAINERENV_TMPDIR is preferred
  INFO:    Environment variable SINGULARITYENV_NXF_DEBUG is set, but APPTAINERENV_NXF_DEBUG is preferred
  INFO:    underlay of /etc/localtime required more than 50 (78) bind mounts
  .command.sh: line 3: seq_cache_populate.pl: command not found
  ps: error while loading shared libraries: librt.so.1: cannot open shared object file: No such file or directory

Work dir:
  /home/groups/ruthm/mthartma/src/ticket/output/workspace/95/0246b881cacf9106560a1df93be122

Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`

 -- Check '.nextflow.log' file for details
WARN: Killing running tasks (1)

[mthartma@sh02-01n58 /home/groups/ruthm/mthartma/src/ticket] (job 35809621) $

**Comment by SamStudio8** [comment]
@mthartma Looks like there is still some mixing between the `projectDir` and the other directories:
```
launchDir : /home/groups/ruthm/mthartma/src/ticket
workDir : /home/groups/ruthm/mthartma/src/ticket/output/workspace
projectDir : /home/users/mthartma/.nextflow/assets/epi2me-labs/wf-human-variation
userName : mthartma
profile : singularity
configFiles : /home/users/mthartma/.nextflow/assets/epi2me-labs/wf-human-variation/nextflow.config
```

What happens if you temporarily move the Nextflow home?:
```
export NXF_HOME=/home/groups/ruthm/mthartma/src/ticket/nxf
```

**Comment by mthartma** [comment]
I'm going to try running this out of my /home/users/mthartma directory first then I'll recreate the /home/groups/ruthm/mthartma version

**Comment by mthartma** [comment]
[mthartma@sh02-01n58 ~/nextflow_test] (job 35811817) $ nextflow run epi2me-labs/wf-human-variation -w ${OUTPUT}/workspace -profile singularity --snp --sv --bam demo_data/demo.bam --bed demo_data/demo.bed --ref demo_data/demo.fasta --basecaller_cfg 'dna_r10.4.1_e8.2_400bps_hac@v4.1.0' --sample_name MY_SAMPLE --out_dir ${OUTPUT}
N E X T F L O W  ~  version 23.04.3
Launching `https://github.com/epi2me-labs/wf-human-variation` [small_mayer] DSL2 - revision: 93a02af3ff [master]

||||||||||   _____ ____ ___ ____  __  __ _____      _       _
||||||||||  | ____|  _ \_ _|___ \|  \/  | ____|    | | __ _| |__  ___
|||||       |  _| | |_) | |  __) | |\/| |  _| _____| |/ _` | '_ \/ __|
|||||       | |___|  __/| | / __/| |  | | |__|_____| | (_| | |_) \__ \
||||||||||  |_____|_|  |___|_____|_|  |_|_____|    |_|\__,_|_.__/|___/
||||||||||  wf-human-variation v1.8.3-g93a02af
--------------------------------------------------------------------------------
Core Nextflow options
  revision       : master
  runName        : small_mayer
  containerEngine: singularity
  container      : ontresearch/wf-human-variation:sha0800eade05e4cbb75d45421633c78c4f6320b2f6
  launchDir      : /home/users/mthartma/nextflow_test
  workDir        : /home/users/mthartma/nextflow_test/output/workspace
  projectDir     : /home/groups/ruthm/mthartma/src/ticket/assets/epi2me-labs/wf-human-variation
  userName       : mthartma
  profile        : singularity
  configFiles    : /home/groups/ruthm/mthartma/src/ticket/assets/epi2me-labs/wf-human-variation/nextflow.config

Workflow Options
  sv             : true
  snp            : true

Main options
  sample_name    : MY_SAMPLE
  bam            : demo_data/demo.bam
  ref            : demo_data/demo.fasta
  basecaller_cfg : dna_r10.4.1_e8.2_400bps_hac@v4.1.0
  bed            : demo_data/demo.bed

!! Only displaying parameters that differ from the pipeline defaults !!
--------------------------------------------------------------------------------
If you use epi2me-labs/wf-human-variation for your analysis please cite:

* The nf-core framework
  https://doi.org/10.1038/s41587-020-0439-x


--------------------------------------------------------------------------------
This is epi2me-labs/wf-human-variation v1.8.3-g93a02af.
--------------------------------------------------------------------------------
executor >  local (1)
executor >  local (1)
executor >  local (2)
[b6/2bad37] process > bam_ingress:check_for_alignment (1) [  0%] 0 of 1
[-        ] process > bam_ingress:cram_to_bam             -
[-        ] process > bam_ingress:minimap2_alignment      -
[-        ] process > getGenome                           -
[-        ] process > cram_cache                          [  0%] 0 of 1
[-        ] process > mosdepth_input                      -
[-        ] process > readStats                           -
[08/ee01ef] process > getVersions                         [  0%] 0 of 1
[-        ] process > getParams                           [  0%] 0 of 1
[-        ] process > get_region_coverage                 -
[-        ] process > makeAlignmentReport                 -
[-        ] process > failedQCReport                      -
[-        ] process > lookup_clair3_model                 -
[-        ] process > snp:make_chunks                     -
[-        ] process > snp:pileup_variants                 -
[-        ] process > snp:aggregate_pileup_variants       -
[-        ] process > snp:select_het_snps                 -
[-        ] process > snp:phase_contig                    -
[-        ] process > snp:get_qual_filter                 -
[-        ] process > snp:create_candidates               -
[-        ] process > snp:evaluate_candidates             -
[-        ] process > snp:aggregate_full_align_variants   -
[-        ] process > snp:merge_pileup_and_full_vars      -
[-        ] process > snp:aggregate_all_variants          -
[-        ] process > sv:variantCall:filterBam            -
[-        ] process > sv:variantCall:sniffles2            -
[-        ] process > sv:variantCall:filterCalls          -
executor >  local (2)
[b6/2bad37] process > bam_ingress:check_for_alignment (1) [100%] 1 of 1, failed: 1 ✘
[-        ] process > bam_ingress:cram_to_bam             -
[-        ] process > bam_ingress:minimap2_alignment      -
[-        ] process > getGenome                           -
[-        ] process > cram_cache                          [  0%] 0 of 1
[-        ] process > mosdepth_input                      -
[-        ] process > readStats                           -
[08/ee01ef] process > getVersions                         [  0%] 0 of 1
[-        ] process > getParams                           [  0%] 0 of 1
[-        ] process > get_region_coverage                 -
[-        ] process > makeAlignmentReport                 -
[-        ] process > failedQCReport                      -
[-        ] process > lookup_clair3_model                 -
[-        ] process > snp:make_chunks                     -
[-        ] process > snp:pileup_variants                 -
[-        ] process > snp:aggregate_pileup_variants       -
[-        ] process > snp:select_het_snps                 -
[-        ] process > snp:phase_contig                    -
[-        ] process > snp:get_qual_filter                 -
[-        ] process > snp:create_candidates               -
[-        ] process > snp:evaluate_candidates             -
[-        ] process > snp:aggregate_full_align_variants   -
[-        ] process > snp:merge_pileup_and_full_vars      -
[-        ] process > snp:aggregate_all_variants          -
[-        ] process > sv:variantCall:filterBam            -
[-        ] process > sv:variantCall:sniffles2            -
[-        ] process > sv:variantCall:filterCalls          -
[-        ] process > sv:variantCall:sortVCF              -
executor >  local (2)
[b6/2bad37] process > bam_ingress:check_for_alignment (1) [100%] 1 of 1, failed: 1 ✘
[-        ] process > bam_ingress:cram_to_bam             -
[-        ] process > bam_ingress:minimap2_alignment      -
[-        ] process > getGenome                           -
[-        ] process > cram_cache                          [  0%] 0 of 1
[-        ] process > mosdepth_input                      -
[-        ] process > readStats                           -
[08/ee01ef] process > getVersions                         [  0%] 0 of 1
[-        ] process > getParams                           [  0%] 0 of 1
[-        ] process > get_region_coverage                 -
[-        ] process > makeAlignmentReport                 -
[-        ] process > failedQCReport                      -
[-        ] process > lookup_clair3_model                 -
[-        ] process > snp:make_chunks                     -
[-        ] process > snp:pileup_variants                 -
[-        ] process > snp:aggregate_pileup_variants       -
[-        ] process > snp:select_het_snps                 -
[-        ] process > snp:phase_contig                    -
[-        ] process > snp:get_qual_filter                 -
[-        ] process > snp:create_candidates               -
[-        ] process > snp:evaluate_candidates             -
[-        ] process > snp:aggregate_full_align_variants   -
[-        ] process > snp:merge_pileup_and_full_vars      -
[-        ] process > snp:aggregate_all_variants          -
[-        ] process > sv:variantCall:filterBam            -
[-        ] process > sv:variantCall:sniffles2            -
[-        ] process > sv:variantCall:filterCalls          -
[-        ] process > sv:variantCall:sortVCF              -
[-        ] process > sv:variantCall:indexVCF             -
[-        ] process > sv:annotate_sv_vcf                  -
[-        ] process > sv:runReport:getVersions            [  0%] 0 of 1
[-        ] process > sv:runReport:getParams              [  0%] 0 of 1
[-        ] process > sv:runReport:report                 -
[-        ] process > output_sv                           [  0%] 0 of 1
[-        ] process > refine_with_sv                      -
[-        ] process > bed_filter                          -
[-        ] process > annotate_snp_vcf                    -
[-        ] process > vcfStats                            -
[-        ] process > report_snp:getVersions              -
[-        ] process > report_snp:getParams                -
[-        ] process > report_snp:makeReport               -
[-        ] process > output_snp                          -
[-        ] process > configure_jbrowse                   -
[-        ] process > publish_artifact                    [  0%] 0 of 1
Pulling Singularity image docker://ontresearch/wf-human-variation-snp:sha0d7e7e8e8207d9d23fdf50a34ceb577da364373e [cache /home/users/mthartma/nextflow_test/nxf/ontresearch-wf-human-variation-snp-sha0d7e7e8e8207d9d23fdf50a34ceb577da364373e.img]
ERROR ~ Error executing process > 'bam_ingress:check_for_alignment (1)'

Caused by:
  Process `bam_ingress:check_for_alignment (1)` terminated with an error exit status (127)

Command executed:

  realign=0
  workflow-glue check_sq_ref --xam demo.bam --ref demo.fasta || realign=$?
  
  # Count SQ lines to help determine if this is unaligned BAM/CRAM later
  # permit grep to exit 1 if the count is zero, otherwise blow up here
  xam_sq_len=$(samtools view -H demo.bam | { grep -c '^@SQ' || [[ $? == 1 ]]; })
  
  # Allow EX_OK and EX_DATAERR, otherwise explode
  if [ $realign -ne 0 ] && [ $realign -ne 65 ]; then
      exit 1
  fi

Command exit status:
  127

Command output:
  (empty)

Command error:
  INFO:    Environment variable SINGULARITYENV_TMPDIR is set, but APPTAINERENV_TMPDIR is preferred
  INFO:    Environment variable SINGULARITYENV_NXF_DEBUG is set, but APPTAINERENV_NXF_DEBUG is preferred
  INFO:    underlay of /etc/localtime required more than 50 (78) bind mounts
  /usr/bin/env: 'python3': No such file or directory
  .command.sh: line 7: samtools: command not found

Work dir:
  /home/users/mthartma/nextflow_test/output/workspace/b6/2bad37aacdc561cf3de233f0db659c

Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`

 -- Check '.nextflow.log' file for details

[mthartma@sh02-01n58 ~/nextflow_test] (job 35811817) $

**Comment by mthartma** [comment]
[mthartma@sh02-01n58 ~/nextflow_test] (job 35811817) $ echo $SINGULARITYENV_NXF_DEBUG 
/home/users/mthartma/nextflow_test/nxf
[mthartma@sh02-01n58 ~/nextflow_test] (job 35811817) $ echo $SINGULARITYENV_TMPDIR 
/home/users/mthartma/nexflow_test
[mthartma@sh02-01n58 ~/nextflow_test] (job 35811817) $ echo $APPTAINERENV_TMPDIR 
/home/users/mthartma/nexflow_test
[mthartma@sh02-01n58 ~/nextflow_test] (job 35811817) $ echo $APPTAINERENV_NXF_DEBUG 
/home/users/mthartma/nextflow_test/nxf
[mthartma@sh02-01n58 ~/nextflow_test] (job 35811817) $

**Comment by mthartma** [comment]
sorry, forgot to update the NXF_HOME variable ...

**Comment by SamStudio8** [comment]
launchDir : /home/users/mthartma/nextflow_test
workDir : /home/users/mthartma/nextflow_test/output/workspace
projectDir : /home/groups/ruthm/mthartma/src/ticket/assets/epi2me-labs/
wf-human-variation
userName : mthartma
profile : singularity
configFiles : /home/groups/ruthm/mthartma/src/ticket/assets/epi2me-labs/
wf-human-variation/nextflow.config

We seem to have the inverse configuration! This should work if you now
launch the workflow from your groups path.

**Comment by mthartma** [comment]
finally got things in order ...

$ export OUTPUT=output
[mthartma@sh02-01n57 ~/nextflow_test] (job 35812379) $ nextflow run epi2me-labs/wf-human-variation -w ${OUTPUT}/workspace -profile singularity --snp --sv --bam demo_data/demo.bam --bed demo_data/demo.bed --ref demo_data/demo.fasta --basecaller_cfg 'dna_r10.4.1_e8.2_400bps_hac@v4.1.0' --sample_name MY_SAMPLE --out_dir ${OUTPUT} 
N E X T F L O W  ~  version 23.04.3
Launching `https://github.com/epi2me-labs/wf-human-variation` [jolly_cantor] DSL2 - revision: 93a02af3ff [master]

||||||||||   _____ ____ ___ ____  __  __ _____      _       _
||||||||||  | ____|  _ \_ _|___ \|  \/  | ____|    | | __ _| |__  ___
|||||       |  _| | |_) | |  __) | |\/| |  _| _____| |/ _` | '_ \/ __|
|||||       | |___|  __/| | / __/| |  | | |__|_____| | (_| | |_) \__ \
||||||||||  |_____|_|  |___|_____|_|  |_|_____|    |_|\__,_|_.__/|___/
||||||||||  wf-human-variation v1.8.3-g93a02af
--------------------------------------------------------------------------------
Core Nextflow options
  revision       : master
  runName        : jolly_cantor
  containerEngine: singularity
  container      : ontresearch/wf-human-variation:sha0800eade05e4cbb75d45421633c78c4f6320b2f6
  launchDir      : /home/users/mthartma/nextflow_test
  workDir        : /home/users/mthartma/nextflow_test/output/workspace
  projectDir     : /home/users/mthartma/nextflow_test/nxf/assets/epi2me-labs/wf-human-variation
  userName       : mthartma
  profile        : singularity
  configFiles    : /home/users/mthartma/nextflow_test/nxf/assets/epi2me-labs/wf-human-variation/nextflow.config

Workflow Options
  sv             : true
  snp            : true

Main options
  sample_name    : MY_SAMPLE
  bam            : demo_data/demo.bam
  ref            : demo_data/demo.fasta
  basecaller_cfg : dna_r10.4.1_e8.2_400bps_hac@v4.1.0
  bed            : demo_data/demo.bed


and it's running.

updating the NXF_HOME variable seems to have worked

**Comment by mthartma** [comment]
ok it ran for a while and then I got this:

ERROR ~ Error executing process > 'snp:pileup_variants (5)'

Caused by:
  Process `snp:pileup_variants (5)` terminated with an error exit status (137)

Command executed:

  python $(which clair3.py) CallVariantsFromCffi             --chkpnt_fn model/pileup             --bam_fn demo.bam             --call_fn pileup_chr20_5.vcf             --ref_fn demo.fasta             --ctgName chr20             --chunk_id 5             --chunk_num 13             --platform ont             --fast_mode False             --snp_min_af 0.08             --indel_min_af 0.15             --minMQ 5             --minCoverage 2             --call_snp_only False             --gvcf false             --temp_file_dir gvcf_tmp_path             --cmd_fn CMD             --pileup             --bed_fn demo.filt.bed --extend_bed split_bed/chr20

Command exit status:
  137

Command output:
  (empty)

Command error:
  INFO:    Environment variable SINGULARITYENV_TMPDIR is set, but APPTAINERENV_TMPDIR is preferred
  INFO:    Environment variable SINGULARITYENV_NXF_DEBUG is set, but APPTAINERENV_NXF_DEBUG is preferred
  INFO:    underlay of /etc/localtime required more than 50 (78) bind mounts
  Calling variants ...
  .command.sh: line 2: 31330 Killed                  python $(which clair3.py) CallVariantsFromCffi --chkpnt_fn model/pileup --bam_fn demo.bam --call_fn pileup_chr20_5.vcf --ref_fn demo.fasta --ctgName chr20 --chunk_id 5 --chunk_num 13 --platform ont --fast_mode False --snp_min_af 0.08 --indel_min_af 0.15 --minMQ 5 --minCoverage 2 --call_snp_only False --gvcf false --temp_file_dir gvcf_tmp_path --cmd_fn CMD --pileup --bed_fn demo.filt.bed --extend_bed split_bed/chr20
  grep: error while loading shared libraries: libdl.so.2: cannot open shared object file: No such file or directory

Work dir:
  /home/users/mthartma/nextflow_test/output/workspace/84/91f69974c60d323cbc19d8126bac8b

Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`

 -- Check '.nextflow.log' file for details
WARN: Killing running tasks (4)

**Comment by SamStudio8** [comment]
@mthartma Oh dear. The good news is that we are indeed making progress. Your workflow is most definitely running. However, an exit 137 implies that one of the processes was killed (probably for using too much memory). It is likely that the workflow is trying to start too many jobs at the same time on your local executor. This will especially be the case for these pileup steps which are run across all the chunks of the data. Tuning workflows is a bit of an art, but to get started you can constrain the number of simultaneous `pileup_variant` jobs by creating a configuration file with the following contents:
```
process.withName:pileup_variants.maxForks = 8
```
This will force Nextflow to run no more than eight pileup_variant tasks simultaneously. You can provide additional configuration files to Nextflow with the `-c` argument. ie. If you named the file `constrain_pileup.config` you would append `-c constrain_pileup.config` to the end of your Nextflow run command.

I'm going to close this issue now, as the reported mystery has been solved. If you continue to run in to trouble or need further guidance on the 137 exit, please don't hesitate to open a new issue.

**Comment by tienpm7723** [comment]
> Hm. I am not _entirely_ sure what to make of this one. It is possible that this error is actually a permissions error masquerading as a command not found error, and the user does not have permissions to run the `python` binary included with the container. Let's try and confirm our suspicions by asking the container for its PATH and whether it can access the bin directory that we ship Python in:
> 
> ```
> singularity exec /home/groups/ruthm/mthartma/src/ticket/ontresearch-wf-human-variation-sv-shabc3ac908a14705f248cdf49f218956ec33e93ef9.img bash -c 'echo $PATH; namei -l /home/epi2melabs/conda/bin/python; which python'
> ```
> 
> @mthartma Before we get too in the weeds, can I just check if you have any options that you generally apply to `singularity exec` when running containers for other tasks? I note you're on CentOS which has required some special configuration to correctly bind mounts for Docker for other users.

Hi im run wf-human-variantion on Rocky Linux 8.9 and when run the above command I get this error:
```
singularity exec /scratch/project/stseq/Tien/test_scripts/singularity/ontresearch-wf-human-variation-sv-sha69e7a43c6f21c0ea2083e9719257fb44bc35c2ba.img bash -c 'echo $PATH; namei -l /home/epi2melabs/conda/bin/python'
/home/epi2melabs/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
f: /home/epi2melabs/conda/bin/python
drwxr-xr-x uqhnguy8 qris-uq /
drwxr-xr-x nobody   nogroup home
                            epi2melabs - No such file or directory
```
May you help me why I cant run the singularity container here? Thank you
